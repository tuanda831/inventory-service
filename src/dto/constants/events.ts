export const EVENT__PRODUCT_CREATION = 'inventory.product.creation';
