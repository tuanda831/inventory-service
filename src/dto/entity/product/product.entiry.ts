import { Column, Entity, PrimaryGeneratedColumn } from 'typeorm';

@Entity({ name: 'products' })
export class Product {
  @PrimaryGeneratedColumn('uuid')
  id: string;

  @Column()
  sku: string;

  @Column()
  shortCode: string;

  @Column()
  barcode: string;

  @Column()
  name: string;

  @Column()
  description: string;

  @Column('decimal')
  price: number;

  @Column({
    type: 'jsonb',
    default: () => "'[]'",
    nullable: false,
  })
  images: string[];

  @Column({
    type: 'timestamptz',
    default: () => 'NOW()::TIMESTAMPTZ',
  })
  createdAt: string;

  @Column({
    type: 'timestamptz',
    default: () => 'NOW()::TIMESTAMPTZ',
  })
  updatedAt: string;

  @Column({
    type: 'timestamptz',
    nullable: true,
  })
  deletedAt: string;

  @Column({
    type: 'timestamptz',
    nullable: true,
  })
  restoredAt: string;
}
