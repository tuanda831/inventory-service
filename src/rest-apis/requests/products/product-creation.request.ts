import {
  IsArray,
  IsNotEmpty,
  IsNumber,
  IsOptional,
  IsString,
} from 'class-validator';

export class ProductCreationRequest {
  @IsString()
  @IsNotEmpty()
  sku: string;

  @IsString()
  @IsNotEmpty()
  barcode: string;

  @IsString()
  @IsNotEmpty()
  name: string;

  @IsString()
  @IsOptional()
  description?: string;

  @IsArray()
  @IsOptional()
  images?: string[];

  @IsNumber()
  @IsNotEmpty()
  price: number;
}
