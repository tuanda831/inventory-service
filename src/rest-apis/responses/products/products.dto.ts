import { ApiProperty, ApiResponseProperty } from '@nestjs/swagger';

export class ProductIdDTO {
  @ApiProperty()
  id: string;
}

export class ProductDTO {
  @ApiProperty()
  id: string;

  @ApiProperty()
  sku: string;

  @ApiProperty()
  shortCode: string;

  @ApiProperty()
  barcode: string;

  @ApiProperty()
  name: string;

  @ApiProperty()
  description: string;

  @ApiProperty()
  price: number;

  @ApiProperty()
  images: string[];

  @ApiResponseProperty({
    example: 'yyyy-MM-ddThh:mm:ss.123Z',
  })
  createdAt: string;

  @ApiResponseProperty({
    example: 'yyyy-MM-ddThh:mm:ss.123Z',
  })
  updatedAt: string;

  @ApiResponseProperty({
    example: 'yyyy-MM-ddThh:mm:ss.123Z',
  })
  deletedAt: string;

  @ApiResponseProperty({
    example: 'yyyy-MM-ddThh:mm:ss.123Z',
  })
  restoredAt: string;
}
