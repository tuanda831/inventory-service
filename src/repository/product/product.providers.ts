import { DataSource, Repository } from 'typeorm';
import { Product } from '../../dto/entity/product/product.entiry';

export const productRepositoryProviders = [
  {
    provide: Repository<Product>,
    useFactory: (dataSource: DataSource) => dataSource.getRepository(Product),
    inject: [DataSource],
  },
];
