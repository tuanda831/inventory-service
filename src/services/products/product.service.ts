import { Inject, Injectable } from '@nestjs/common';
import { Product } from '../../dto/entity/product/product.entiry';
import { SearchServicePublisher } from '../../workflows/search-service/publisher';
import { Repository } from 'typeorm';

@Injectable()
export class ProductService {
  constructor(
    @Inject(Repository<Product>)
    private productRepository: Repository<Product>,
    @Inject(SearchServicePublisher)
    private searchPublisher: SearchServicePublisher,
  ) {}

  gets(): Promise<Product[]> {
    return this.productRepository.find();
  }

  async save(product: Product): Promise<Product> {
    await this.searchPublisher.addProductToIndex(product);
    return this.productRepository.save(product);
  }
}
