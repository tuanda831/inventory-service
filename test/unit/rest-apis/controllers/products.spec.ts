import { ClientKafka } from '@nestjs/microservices';
import { Response } from 'express';
import { Observable } from 'rxjs';
import { Repository } from 'typeorm';
import { EVENT__PRODUCT_CREATION } from '../../../../src/dto/constants/events';
import { Product } from '../../../../src/dto/entity/product/product.entiry';
import { ProductController } from '../../../../src/rest-apis/controllers/products.controller';
import { ProductCreationRequest } from '../../../../src/rest-apis/requests/products/product-creation.request';
import { ProductService } from '../../../../src/services/products/product.service';

let controller: ProductController;
let productService: ProductService;
let productRepository: Repository<Product>;
let eventClient: ClientKafka;

describe('ProductController', () => {
  beforeEach(async () => {
    productService = new ProductService(productRepository);
    eventClient = new ClientKafka({});
    controller = new ProductController(productService, eventClient);
  });

  describe('/products (POST)', () => {
    const request: ProductCreationRequest = {
      sku: '123',
      barcode: '99999',
      name: 'New Product',
      description: 'Product Description here',
      images: ['https://test1.link', 'https://test2.link'],
      price: 1.98,
    };

    let res: Response;

    it('Should Return 400 incase the request invalid', () => {
      expect(true).toBe(true);
    });

    it('Should handle the Unknow error well', () => {
      expect(true).toBe(true);
    });

    it('Should successful insert a record to the DB', async () => {
      jest
        .spyOn(eventClient, 'emit')
        .mockImplementation(
          (eventName: string, products: Product[]): Observable<Product> => {
            expect(eventName).toBe(EVENT__PRODUCT_CREATION);
            expect(products[0]).toMatchSnapshot({
              id: expect.any(String),
            });
            return new Observable<Product>();
          },
        );

      res = {
        status: (code: any): Response => {
          expect(code).toBe(201);
          return {
            send({ message, data }) {
              expect(message).toBe('"Product" is Created!');
              expect(data).toMatchObject({
                id: expect.any(String),
              });
            },
          } as Response;
        },
      } as Response;

      await controller.create(res, request);
    });
  });
});
