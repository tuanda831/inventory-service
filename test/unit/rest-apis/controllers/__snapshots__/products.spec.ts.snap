// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductController /products (POST) Should successful insert a record to the DB 1`] = `
Object {
  "barcode": "99999",
  "description": "Product Description here",
  "id": Any<String>,
  "images": Array [
    "https://test1.link",
    "https://test2.link",
  ],
  "name": "New Product",
  "price": 1.98,
  "shortCode": "New-Product-123",
  "sku": "123",
}
`;
